### PR TITLE
[Apple Silicon] Add an architecture-independent ATen-to-MLX lowering layer

### PR DESCRIPTION
**Pull request by noob-se7en**
## Motivation

Split out of #36164 (the MLX whole-region serving PR) so that PR stays reviewable and this layer can be reviewed on its own. Review there flagged the ATen-to-MLX mapping as hard-coded in two places (a mapping table plus a 34-branch dispatch chain) and suggested a config file; this PR lands the layer in its final shape instead.

This module is self-contained: it classifies an exported/compiled FX graph by operation rather than model architecture and lowers admitted ATen ops to MLX. No serving state — no KV pools, no attention contract, no runners. #36164 rebases onto it and keeps only the serving executor.

## Modifications

- `hardware_backend/mlx/fx_lowering.py`:
  - Each lowering is one function registered with `@_lowering(name, aten=..., functions=..., methods=...)`; both standard registries are derived from those registrations. Adding an op is a single self-contained function — no table edit, no new branch, and no stringly-typed config entry that can typo silently.
  - 35 op lowerings, FX graph planning (`build_mlx_fx_plan`), a generic single-region MLX executor over borrowed Torch tensors, export planning (`export_mlx_plan`, functionalized with explicit user-input mutations), and the Dynamo validation backend (`MlxFxCaptureBackend`) reporting whole-graph lowering coverage.

## Accuracy Tests

`test_fx_lowering.py` (11 tests; planning on CPU, execution on Metal): whole-graph capture of a dense decoder with a KV-commit custom op; executor parity against eager Torch in one region; one Torch graph reused across token counts; weight mutation/replacement observed through borrowed views; real SGLang MLP architectures (incl. Qwen3) with no per-architecture registry; functionalized KV mutation surfaced as an explicit graph output; whole-region rejection with eager Torch fallback for unsupported ops.

## How torch/MLX version bumps are re-verified

The schema parity suite (`test_fx_lowering_schema_parity.py`) reads both libraries live on every run, nothing is snapshotted. The core check: for every op we lower, the test runs the op twice with identical inputs, once in plain torch and once through the MLX lowering, and the two results must agree.

**Torch bump**
- Test cases are generated from each op's `_schema` read from the installed torch at collection (base case plus one case per defaulted argument with the default overridden).
- A default changed: plain torch now computes with the new default. If our lowering assumed the old one, the results disagree and the test fails.
- A new argument with a default appeared: the suite reads it from the signature and tests it automatically.
- An op we lower was removed or renamed: our registration can't find it and the module fails to import.
- torch.export starts putting a new op into real model graphs: the end-to-end tests (a real Qwen3 MLP exported and executed) reject the graph and say which op has no lowering.

**MLX bump**
- Every test calls the installed MLX functions for real (`mx.fast.rms_norm`, `mx.fast.scaled_dot_product_attention`, and so on). If MLX changed a function's signature, the call throws and the test fails. If MLX changed the numbers it produces, they no longer match torch and the test fails.

This suite caught two keyword-blind lowerings on its first run (flatten and slice dropping keyword arguments, the same class as the alpha bug from review), fixed in 6b1672a and 3cc5e59.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #35200822679](https://github.com/sgl-project/sglang/actions/runs/35200822679)
Latest PR Test (Extra): :x: [Run #35200822402](https://github.com/sgl-project/sglang/actions/runs/35200822402)
Latest PR Test (AMD ROCm 10): :x: [Run #35200822575](https://github.com/sgl-project/sglang/actions/runs/35200822575)

### CODE REVIEW

**Comment by noob-se7en** [comment]
cc: @yeahdongcn

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py [hunk 1]
```diff
@@ -0,0 +1,745 @@
...
+@_lowering(
+    "linear",
+    aten=(torch.ops.aten.linear.default, torch.ops.aten.mm.default),
+    functions=(torch._C._nn.linear,),
+)
+def _lower_linear(mx, args, kwargs):
+    value, weight = args[:2]
+    bias = _arg(args, kwargs, 2, "bias")
+    output = value @ mx.swapaxes(weight, -1, -2)
+    return output if bias is None else output + bias
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/fx_lowering.py:272]: Is this correct? `aten.mm(a, b)` means `a @ b`.

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py [hunk 2]
```diff
@@ -0,0 +1,745 @@
...
+
+
+@_lowering("multiply", aten=(torch.ops.aten.mul.Tensor,), functions=(operator.mul,))
+def _lower_multiply(mx, args, kwargs):
+    return args[0] * args[1]
+
+
+@_lowering("subtract", aten=(torch.ops.aten.sub.Tensor,))
+def _lower_subtract(mx, args, kwargs):
+    return args[0] - args[1] * kwargs.get("alpha", 1)
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/fx_lowering.py]: From codex: The ATen overload `aten.sub.Tensor(self, other, alpha=1)` may carry `alpha` as the third positional argument in an exported graph. This implementation only reads `kwargs["alpha"]`, so a graph containing `x - 2 * y` can be lowered as `x - y` without any error. Read alpha from the same positional/keyword helper used elsewhere (and reject unsupported non-scalar alpha if necessary). `add.Tensor` has the analogous alpha contract and should be audited as well.

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py [hunk 3]
```diff
@@ -0,0 +1,745 @@
...
+    attr_nodes = tuple(
+        node for node in plan.graph_module.graph.nodes if node.op == "get_attr"
+    )
+    attr_views = tuple(
+        (
+            MlxTensorView(plan.graph_module.get_parameter(str(node.target)))
+            if str(node.target) in dict(plan.graph_module.named_parameters())
+            else MlxTensorView(plan.graph_module.get_buffer(str(node.target)))
+        )
+        for node in attr_nodes
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/fx_lowering.py]: From codex: `get_attr` can refer to a plain module attribute (for example a scalar, tuple, or constant tensor), not only a parameter or registered buffer. The constructor calls `get_buffer()` for every attribute that is not found in `named_parameters()`, which raises before execution for such valid FX graphs. Resolve attributes with a three-way path: parameter, buffer, then ordinary attribute; convert tensor attributes to `MlxTensorView` and pass immutable scalar/shape metadata as captured values. Add a test for a module containing `self.scale = 2.0` and a tensor constant.

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+    stop = args[3] if len(args) > 3 else value.shape[axis]
+    step = args[4] if len(args) > 4 else 1
+    width = value.shape[axis]
+    if start >= 0:
+        start = min(start, width)
+    if stop >= 0:
+        stop = min(stop, width)
+    slices = [slice(None)] * value.ndim
+    slices[axis] = slice(start, stop, step)
+    return value[tuple(slices)]
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/fx_lowering.py:502]: From codex: `aten.slice.Tensor` uses `None`/`INT64_MAX` and normalizes negative indices according to Python/PyTorch rules. The current code leaves a negative `stop` unchanged, so `stop=-1` drops the last element whereas PyTorch's common `[..., :-1]` and exported sentinel cases require different normalization. Please centralize bound normalization (including negative start/stop and negative step, or explicitly reject unsupported steps) and add parity tests.

**Review by yeahdongcn — COMMENTED** [review]
I think maintaining such a large mapping would be difficult. Moreover, whenever we bump the PyTorch or MLX versions, it would require substantial additional work and verification. Is there a way to generate this mapping instead? We could define a skeleton and build the mapping from a configuration file that also specifies the pinned PyTorch and MLX versions. (Yes, AI agents make this much easier)

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py (hunk 1 above)
Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/fx_lowering.py:272]: Good catch, this was wrong. mm routed through the linear lowering, which transposes the second operand, so a bare `mm(a, b)` computed `a @ b.T`. It was masked because the serving path doesn't decompose (nn.Linear stays linear.default and .T arrives as numpy_T + matmul), so no real graph hit it. Now routed to the plain matmul lowering, with a parity test: `6b1672a`

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py (hunk 2 above)
Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/fx_lowering.py]: Half right. alpha is keyword-only in the schema, so the positional case can't occur and sub's kwargs read was fine. The real bug was next door: add dropped alpha entirely, so `add(x, y, alpha=2)` computed `x + y`. Both now go through the positional/keyword helper, with alpha != 1 parity tests: `6b1672a`

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py (hunk 3 above)
Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/fx_lowering.py]: Correct. Only parameters and buffers resolved, so a constant tensor hung directly on the module crashed executor construction. Resolution is now parameter, then buffer, then plain attribute, with tensors as borrowed views and scalars captured by value. Test added: 6b1672a

File: python/sglang/srt/hardware_backend/mlx/fx_lowering.py (hunk 4 above)
Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/fx_lowering.py:502]: Added parity tests for the negative stop, sentinel, and negative start cases. They pass unchanged: torch's slice normalization already matches Python/MLX slicing here, so this just pins the behaviour with a test: 6b1672a

**Comment by noob-se7en** [comment]
> I think maintaining such a large mapping would be difficult. Moreover, whenever we bump the PyTorch or MLX versions, it would require substantial additional work and verification. Is there a way to generate this mapping instead? We could define a skeleton and build the mapping from a configuration file that also specifies the pinned PyTorch and MLX versions. (Yes, AI agents make this much easier)

Discussing this in detail on slack, will update the final decision here.

**Comment by noob-se7en** [comment]
> > I think maintaining such a large mapping would be difficult. Moreover, whenever we bump the PyTorch or MLX versions, it would require substantial additional work and verification. Is there a way to generate this mapping instead? We could define a skeleton and build the mapping from a configuration file that also specifies the pinned PyTorch and MLX versions. (Yes, AI agents make this much easier)
> 
> Discussing this in detail on slack, will update the final decision here.

Summary of the discussion:

The core check is simple: for every op we lower, the test runs the op twice with identical inputs, once in plain torch and once through our MLX lowering, and the two results must agree.

For torch bump:

Test cases are generated from each op's _schema read from the installed torch at collection (base case plus one case per defaulted argument with the default overridden).
If a default changed: plain torch now computes with the new default. If our lowering assumed the old one, the results disagree and the test fails.
A new argument with a default appeared: the suite reads it from the signature and tests it automatically. 
An op we lower was removed or renamed: our registration can't find it and the module fails to import.
torch.export starts putting a new op into real model graphs: the end-to-end tests (a real Qwen3 MLP exported and executed) reject the graph and say which op has no lowering.


For MLX Bump:

Every test calls the installed MLX functions for real (mx.fast.rms_norm , mx.fast.scaled_dot_product_attention, and so on). If MLX changed a function's signature, the call throws and the test fails. If MLX changed the numbers it produces, they no longer match torch and the test fails.

**Review by yeahdongcn — APPROVED** [review]

**Comment by yeahdongcn** [comment]
@alexnails MLX CI passed. Could you take another look? Thanks!
